sky/serve/autoscalers.py

```diff
@@ -192,5 +196,196 @@ def evaluate_scaling(
+        all_replica_ids_to_scale_down: List[int] = []
+
+        def _get_replica_ids_to_scale_down(
+            info_filter: Callable[['replica_managers.ReplicaInfo'], bool],
+            status_order: List['serve_state.ReplicaStatus'],
+            num_limit: int,
+        ) -> List[int]:
+            replica_ids_to_scale_down: List[int] = []
+            for target_status in status_order:
+                for info in alive_replica_infos:
+                    if info_filter(info) and info.status == target_status:
+                        if len(replica_ids_to_scale_down) >= num_limit:
+                            return replica_ids_to_scale_down
+                        replica_ids_to_scale_down.append(info.replica_id)
+            for info in alive_replica_infos:
+                if info_filter(info) and info.status not in status_order:
+                    if len(replica_ids_to_scale_down) >= num_limit:
+                        return replica_ids_to_scale_down
+                    replica_ids_to_scale_down.append(info.replica_id)
+            return replica_ids_to_scale_down
+
+        num_to_provision = (self.target_num_replicas +
+                            _DEFAULT_OVER_PROVISION_NUM)
+
+        # Scale spot instances.
+        if num_alive_spot < num_to_provision:
+            # Not enough spot instances, scale up.
+            num_spot_to_scale_up = num_to_provision - num_alive_spot
+            for _ in range(num_spot_to_scale_up):
+                spot_override = self._get_spot_resources_override_dict()
+                zone = self.spot_placer.select()
+                spot_override.update({'zone': zone})
+                logger.info(f'Chosen zone {zone} with {self.spot_placer}')
+                scaling_options.append(
+                    AutoscalerDecision(AutoscalerDecisionOperator.SCALE_UP,
+                                       target=spot_override))
+        elif num_alive_spot > num_to_provision:
+            # Too many spot instances, scale down.
+            num_spot_to_scale_down = num_alive_spot - num_to_provision
+            all_replica_ids_to_scale_down.extend(
+                _get_replica_ids_to_scale_down(
+                    info_filter=lambda info: info.is_spot,
+                    status_order=serve_state.ReplicaStatus.
+                    scale_down_decision_order(),
+                    num_limit=num_spot_to_scale_down,
+                ))
+
+        # OnDemand fallback.
+        if num_ready_spot + num_on_demand < num_to_provision:
+            # Enable OnDemand fallback.
+            num_on_demand_to_scale_up = min(
+                self.target_num_replicas,
+                num_to_provision - num_ready_spot) - num_on_demand
+            for _ in range(num_on_demand_to_scale_up):
+                scaling_options.append(
+                    AutoscalerDecision(
+                        AutoscalerDecisionOperator.SCALE_UP,
+                        target=self._get_on_demand_resources_override_dict()))
```

Comment:
nit: should we also do something like
`on_demand_override.update({'zone': zone})`?

Reply:
If we assume that spot and on-demand have some correlation (maybe not)

Reply:
This is to avoid any quota/availability issues. If we don't set zones then we could failover to all enabled zones